crates/compiler/src/visitors/string_table_generator_visitor.rs

```diff
@@ -0,0 +1,236 @@
+                StringInfo {
+                    text: ctx.body().unwrap().get_text(),
+                    node_name: self.current_node_name.clone(),
+                    line_number: ctx.body().unwrap().start().line as usize,
+                    file_name: self.file_name.clone(),
+                    ..Default::default()
+                },
+            );
+        } else {
+            // This is a regular node
+            // String table generator: don't crash if a node has no body
+            if let Some(body) = ctx.body() {
+                self.visit(&*body);
+            }
+        }
+    }
+
+    fn visit_line_statement(&mut self, ctx: &Line_statementContext<'input>) -> Self::Return {
+        let hashtags = ctx.hashtag_all();
+        let line_id_tag = get_line_id_tag(&hashtags);
+        let line_id = line_id_tag.as_ref().and_then(|t| t.text.as_ref());
+
+        if let Some(line_id) = line_id {
+            if self.string_table_manager.contains_key(&line_id.to_string()) {
+                // The original has a fallback for when this is `null` / `None`,
+                // but this can logically not be the case in this scope.
+                let diagnostic_context = line_id_tag.clone().unwrap();
+                let line_id = line_id.get_text();
+                self.diagnostics.push(
+                    Diagnostic::from_message(format!("Duplicate line ID {line_id}"))
+                        .read_parser_rule_context(diagnostic_context)
+                        .with_file_name(&self.file_name),
+                );
+                return;
+            }
+        };
+
+        let line_number = ctx.start().line;
+        let hashtag_texts = get_hashtag_texts(&hashtags);
+
+        let composed_string = generate_formatted_text(&ctx.line_formatted_text().unwrap());
+
+        let string_id = self.string_table_manager.insert(
+            line_id.map(|t| t.get_text().to_owned()),
+            StringInfo {
+                text: composed_string,
+                node_name: self.current_node_name.clone(),
+                line_number: line_number as usize,
+                file_name: self.file_name.clone(),
+                metadata: hashtag_texts,
+                ..Default::default()
+            },
+        );
+
+        if line_id.is_none() {
+            add_hashtag_child(ctx, string_id);
+        }
+    }
+}
+
+/// Takes a string like
+/// `Hi there { some_expression }, how are you { another_expression } doing?`
+/// and turns it into
+/// `Hi there {0}, how are you {1}? doing`
+fn generate_formatted_text(ctx: &Line_formatted_textContext) -> String {
+    let mut expression_count = 0;
+    let mut composed_string = String::new();
+    // First, visit all of the nodes, which are either terminal
+    // text nodes or expressions. if they're expressions, we
+    // evaluate them, and inject a positional reference into the
+    // final string.
+    for child in ctx.get_children() {
+        if child.get_child_count() == 0 {
```

Comment:
is there a more efficient empty method or is that constant time?

Reply:
The API does not provide anything. There's an `is_empty`, but it doesn't do what you'd think. I could do `get_children().next().is_empty()`, but that allocates a `Box`, so its probs more expensive then whatever this does. I bet it's constant time tho.